### ISSUE TITLE
this is a test

### ISSUE THREAD

**Issue by xinqinglxl** [xinqinglxl/FileKeeper#1]
test something

**Comment by xinqinglxl** [comment]
from sklearn.externals import joblib
from keras.preprocessing.sequence import pad_sequences
from keras.utils import to_categorical
import numpy as np

tokenizer = joblib.load('cnn_tokenizer_300d_maxlen2048_wordLenAbove2_totalTokenizer_0509.pkl')

MAX_SEQUENCE_LENGTH = 2048
EMBEDDING_DIM = 300

train_sequences = tokenizer.texts_to_sequences(train_texts)
x_train = pad_sequences(train_sequences, maxlen=MAX_SEQUENCE_LENGTH)
y_train = to_categorical(np.asarray(train_labels))
print('Shape of train data tensor:', x_train.shape)
print('Shape of train label tensor:', y_train.shape)

test_sequences = tokenizer.texts_to_sequences(test_texts)
x_test = pad_sequences(test_sequences, maxlen=MAX_SEQUENCE_LENGTH)
y_test = to_categorical(np.asarray(test_labels))

word_index = tokenizer.word_index

**Comment by xinqinglxl** [comment]
MAX_SEQUENCE_LENGTH = 2048
EMBEDDING_DIM = 300

# sep_index = len(train_texts)
# all_labels = train_labels + test_labels
# all_texts = train_texts + test_texts



train_sequences = tokenizer.texts_to_sequences(train_texts)
x_train = pad_sequences(train_sequences, maxlen=MAX_SEQUENCE_LENGTH)
y_train = to_categorical(np.asarray(train_labels))
print('Shape of train data tensor:', x_train.shape)
print('Shape of train label tensor:', y_train.shape)

test_sequences = tokenizer.texts_to_sequences(test_texts)
x_test = pad_sequences(test_sequences, maxlen=MAX_SEQUENCE_LENGTH)
y_test = to_categorical(np.asarray(test_labels))
print('Shape of test data tensor:', x_test.shape)
print('Shape of test label tensor:', y_test.shape)

**Comment by xinqinglxl** [comment]
inputs = Input(shape=(MAX_SEQUENCE_LENGTH,), dtype='int32', name="cnn1_input")
embedding = Embedding(input_dim=len(word_index) + 1, output_dim=EMBEDDING_DIM, input_length=MAX_SEQUENCE_LENGTH)(inputs)
dropout = Dropout(DROP_RATIO)(embedding)
conv1 = Conv1D(FILTER_NUM, KERNEL_SIZES, padding='valid', activation='relu', strides=1)(dropout)
maxpool = MaxPooling1D(MAX_POOL_SIZE)(conv1)
flatten = Flatten()(maxpool)
dense1 = Dense(EMBEDDING_DIM, activation='relu')(flatten)
output = Dense(y_train.shape[1], activation='softmax')(dense1)

model1 = Model(inputs=inputs, outputs=output)
model1.summary()

checkpoint = ModelCheckpoint('cnn_model1.{epoch:03d}-{val_acc:.4f}.hdf5', monitor='val_acc', verbose=1, save_best_only=True, mode='auto')
model1.compile(loss='categorical_crossentropy', optimizer="adam", metrics=['acc'])
print(model1.metrics_names)
train_his = model1.fit(x_train, y_train, validation_data=(x_test, y_test), callbacks=[checkpoint], epochs=EPOCH_NUM, batch_size=BATCH_SIZE)

**Comment by xinqinglxl** [comment]
from sklearn.metrics import confusion_matrix
import matplotlib.pyplot as plt
import numpy as np

%matplotlib inline

y_preds = np.argmax(preds,axis=1)
y_true = np.argmax(y_test, axis = 1)
cm = confusion_matrix(y_true = y_true, y_pred = y_preds)
plt.imshow(cm, interpolation="nearest", cmap=plt.cm.Blues)

# make various adjustments to the plot
plt.tight_layout()
plt.colorbar()
tick_marks = np.arange(len(cls_labels))
plt.xticks(tick_marks, range(len(cls_labels)))
plt.yticks(tick_marks, range(len(cls_labels)))
plt.xlabel("Predicted")
plt.ylabel(("True"))
plt.show()

**Comment by xinqinglxl** [comment]
inputs = Input(shape=(MAX_SEQUENCE_LENGTH,), dtype='int32', name="cnn2_input")
embedding = Embedding(input_dim=len(word_index) + 1, output_dim=EMBEDDING_DIM, input_length=MAX_SEQUENCE_LENGTH)(inputs)
dropout = Dropout(DROP_RATIO)(embedding)
conv1 = Conv1D(FILTER_NUM, KERNEL_SIZES, padding='valid', activation='relu', strides=1)(dropout)
maxpool = GlobalMaxPooling1D()(conv1)
dense1 = Dense(EMBEDDING_DIM, activation='relu')(maxpool)
dropout2 = Dropout(DROP_RATIO)(dense1)
output = Dense(y_train.shape[1], activation='softmax')(dropout2)

model2 = Model(inputs=inputs, outputs=output)
model2.summary()

checkpoint = ModelCheckpoint('cnn_model2.{epoch:03d}-{val_acc:.4f}.hdf5', monitor='val_acc', verbose=1, save_best_only=True, mode='auto')
model2.compile(loss='categorical_crossentropy', optimizer="adam", metrics=['acc'])
print(model2.metrics_names)
train_his = model2.fit(x_train, y_train, validation_data=(x_test, y_test), callbacks=[checkpoint], epochs=EPOCH_NUM, batch_size=BATCH_SIZE)

**Comment by xinqinglxl** [comment]
def create_channel(input_tensor, filter_size, output_channels):
    """
    Creates a layer, working channel wise
    
    Arguments:
        x           : Input for convoltuional channel
        filter_size : Filter size for creating Conv1D
        output_channels : The number of output filters in the convolution
        
    Returns:
        channel           : Channel including (Conv1D + GlobalMaxPooling + Dense + Dropout)
    """
    channel = SeparableConv1D(output_channels, kernel_size=filter_size, activation='relu', strides=1, padding='same',
                        depth_multiplier=4)(input_tensor)
    channel = GlobalMaxPooling1D()(channel)
    channel = Dense(FC_HIDDEN_NUM)(channel)
    channel = Dropout(DROPOUT_RATIO)(channel)
    return channel

**Comment by xinqinglxl** [comment]
word_input = Input(shape=(MAX_SEQUENCE_LENGTH,), dtype='int32', name='word_input')
embedding_layer = Embedding(len(word_index) + 1, EMBEDDING_DIM, input_length=MAX_SEQUENCE_LENGTH)
input_tensor = embedding_layer(word_input)
input_tensor = Dropout(DROPOUT_RATIO)(input_tensor)
#Add three blocks in parallel, each contains SeparableConv1D + GlobalMaxPooling1D + Dense + Dropout layers
channels = []
for i in range(len(KERNEL_SIZES)):
    x = create_channel(input_tensor, KERNEL_SIZES[i], OUTPUT_NUM)
    channels.append(x)
x = concatenate(channels)
x = Activation('relu')(x)
prediction = Dense(5, activation='softmax')(x)
model3 = Model(inputs=word_input, outputs=prediction)

checkpoint = ModelCheckpoint('cnn_model3.{epoch:03d}-{val_acc:.4f}.hdf5', monitor='val_acc', verbose=1, save_best_only=True, mode='auto')
model3.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])
train_his = model3.fit(x_train, y_train,
        batch_size=128,
        epochs=20,
        callbacks=[checkpoint],
        validation_data=(x_test, y_test))

**Comment by xinqinglxl** [comment]
inputs = Input(shape=(MAX_SEQUENCE_LENGTH,), dtype='int32')
embedding = Embedding(input_dim=len(word_index) + 1, output_dim=EMBEDDING_DIM, input_length=MAX_SEQUENCE_LENGTH)(inputs)
reshape = Reshape((MAX_SEQUENCE_LENGTH,EMBEDDING_DIM,1))(embedding)

#since kernel size is set as (filter_size, embedding_dim), the conv2d works as conv1d
conv_0 = Conv2D(FILTER_NUM, kernel_size=(FILTER_SIZE[0], EMBEDDING_DIM), padding='valid', kernel_initializer='normal', activation='relu')(reshape)
conv_1 = Conv2D(FILTER_NUM, kernel_size=(FILTER_SIZE[1], EMBEDDING_DIM), padding='valid', kernel_initializer='normal', activation='relu')(reshape)
conv_2 = Conv2D(FILTER_NUM, kernel_size=(FILTER_SIZE[2], EMBEDDING_DIM), padding='valid', kernel_initializer='normal', activation='relu')(reshape)

maxpool_0 = MaxPool2D(pool_size=(MAX_SEQUENCE_LENGTH - FILTER_SIZE[0] + 1, 1), strides=(1,1), padding='valid')(conv_0)
maxpool_1 = MaxPool2D(pool_size=(MAX_SEQUENCE_LENGTH - FILTER_SIZE[1] + 1, 1), strides=(1,1), padding='valid')(conv_1)
maxpool_2 = MaxPool2D(pool_size=(MAX_SEQUENCE_LENGTH - FILTER_SIZE[2] + 1, 1), strides=(1,1), padding='valid')(conv_2)

concatenated_tensor = Concatenate(axis=1)([maxpool_0, maxpool_1, maxpool_2])
flatten = Flatten()(concatenated_tensor)
dropout = Dropout(DROP_RATIO)(flatten)
output = Dense(units=y_train.shape[1], activation='softmax')(dropout)

model5 = Model(inputs=inputs, outputs=output)

checkpoint = ModelCheckpoint('cnn_model5.{epoch:03d}-{val_acc:.4f}.hdf5', monitor='val_acc', verbose=1, save_best_only=True, mode='auto')
# adam = Adam(lr=1e-3, beta_1=0.9, beta_2=0.999, epsilon=1e-08, decay=0.2)

model5.compile(optimizer="adam", loss='categorical_crossentropy', metrics=['accuracy'])
train_his = model5.fit(x_train, y_train, batch_size=BATCH_SIZE, epochs=EPOCH_NUM, verbose=1, callbacks=[checkpoint], validation_data=(x_test, y_test))

**Comment by xinqinglxl** [comment]
import matplotlib.pyplot as plt
%matplotlib inline

hist = train_his

plt.figure(1)

axes = plt.gca()
x_min = hist.epoch[0]
x_max = hist.epoch[-1]+1
axes.set_xlim([x_min,x_max])

plt.scatter(hist.epoch, hist.history['loss'], color='g')
plt.plot(hist.history['loss'], color='g', label='Training Loss')
plt.scatter(hist.epoch, hist.history['val_loss'], color='b')
plt.plot(hist.history['val_loss'], color='b', label='Validation Loss')
plt.xlabel('epochs')
plt.ylabel('Loss')
plt.title('Training Loss & Validation Loss vs Epochs')
plt.legend()

plt.figure(2)

axes = plt.gca()
x_min = hist.epoch[0]
x_max = hist.epoch[-1]+1
axes.set_xlim([x_min,x_max])

plt.scatter(hist.epoch, hist.history['acc'], color='r')
plt.plot(hist.history['acc'], color='r', label='Training Accuracy')
plt.scatter(hist.epoch, hist.history['val_acc'], color='c')
plt.plot(hist.history['val_acc'], color='c', label='Validation Accuracy')
plt.xlabel('epochs')
plt.ylabel('Accuracy')
plt.title('Trainging Accuracy & Validation Accuracy vs Epochs')
plt.legend()

plt.show()
